File: docs/pcsi/Evaluations/DS/DS2/permutation.py
def insere(lst, elt):
    return [lst[:i] + [elt] + lst[i:] for i in range(len(lst)+1)]

def permutation(lst):
    if len(lst) == 0:
        return [[]]
    temp =  permutation(lst[1:])
    res = []
    for x in temp:
        res += insere(x, lst[0])
    return res

File: mp2i/files/C6/permutation.ml
let rec insere elt liste pos =
  if pos = 0 then elt::liste else
     (List.hd liste)::(insere elt (List.tl liste) (pos-1));;

let rec insere_all_aux elt liste pos acc =
  if pos=((List.length liste)+1) then acc else insere_all_aux elt liste (pos+1) ((insere elt liste pos)::acc);;

let rec insere_all elt lliste =
  match lliste with
  | [] -> []
  | h::t -> (insere_all_aux elt h 0 [])@(insere_all elt t);;

let rec permutation l =
  match l with
  | [] -> [[]]
  | h::t -> insere_all h (permutation t);;